File: packages/contracts/src/letter.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const letterRequestStatusSchema = z.enum(["SUBMITTED", "APPROVED", "REJECTED", "READY"]);

export type LetterRequestStatus = z.infer<typeof letterRequestStatusSchema>;

export const letterTypeSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1),
  description: z.string().nullable(),
});

export type LetterType = z.infer<typeof letterTypeSchema>;

export const letterTypeListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(letterTypeSchema) }),
);

export type LetterTypeListResponse = z.infer<typeof letterTypeListResponseSchema>;

export const createLetterRequestInputSchema = z
  .object({
    letterTypeId: z.string().uuid(),
    purpose: z.string().trim().min(3).max(1000),
  })
  .strict();

export type CreateLetterRequestInput = z.infer<typeof createLetterRequestInputSchema>;

export const rejectLetterRequestInputSchema = z
  .object({
    reason: z.string().trim().min(3).max(500),
  })
  .strict();

export type RejectLetterRequestInput = z.infer<typeof rejectLetterRequestInputSchema>;

export const letterRequestListQuerySchema = z.object({
  status: letterRequestStatusSchema.optional(),
  limit: z.coerce.number().int().min(1).max(50).default(20),
});

export const letterRequestSchema = z.object({
  id: z.string().uuid(),
  letterTypeId: z.string().uuid(),
  letterTypeName: z.string().min(1),
  purpose: z.string().min(1),
  status: letterRequestStatusSchema,
  requesterName: z.string().min(1),
  houseCode: z.string().min(1),
  householdDisplayName: z.string().min(1),
  reviewedByName: z.string().nullable(),
  reviewedAt: z.string().datetime({ offset: true }).nullable(),
  rejectionReason: z.string().nullable(),
  readyAt: z.string().datetime({ offset: true }).nullable(),
  createdAt: z.string().datetime({ offset: true }),
});

export type LetterRequest = z.infer<typeof letterRequestSchema>;

export const letterRequestListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(letterRequestSchema) }),
);

export const letterRequestMutationResponseSchema = dataEnvelopeSchema(
  z.object({ request: letterRequestSchema }),
);

export type LetterRequestListResponse = z.infer<typeof letterRequestListResponseSchema>;
export type LetterRequestMutationResponse = z.infer<typeof letterRequestMutationResponseSchema>;

File: apps/mobile/lib/features/letter/domain/letter.dart
enum LetterRequestStatus { submitted, approved, rejected, ready }

LetterRequestStatus letterRequestStatusFromApi(Object? value) {
  switch (value) {
    case 'APPROVED':
      return LetterRequestStatus.approved;
    case 'REJECTED':
      return LetterRequestStatus.rejected;
    case 'READY':
      return LetterRequestStatus.ready;
    default:
      return LetterRequestStatus.submitted;
  }
}

const Map<LetterRequestStatus, String> letterRequestStatusLabels = {
  LetterRequestStatus.submitted: 'Diajukan',
  LetterRequestStatus.approved: 'Disetujui',
  LetterRequestStatus.rejected: 'Ditolak',
  LetterRequestStatus.ready: 'Dokumen siap',
};

/// Rough visual weight for a status, mirroring the web app's
/// `statusBadgeVariant` so a mobile status badge reads the same way.
enum LetterStatusTone { muted, warning, success, danger }

LetterStatusTone letterRequestStatusTone(LetterRequestStatus status) {
  switch (status) {
    case LetterRequestStatus.submitted:
      return LetterStatusTone.muted;
    case LetterRequestStatus.approved:
      return LetterStatusTone.warning;
    case LetterRequestStatus.ready:
      return LetterStatusTone.success;
    case LetterRequestStatus.rejected:
      return LetterStatusTone.danger;
  }
}

class LetterType {
  const LetterType({
    required this.id,
    required this.name,
    required this.description,
  });

  final String id;
  final String name;
  final String? description;

  factory LetterType.fromJson(Map<String, dynamic> json) {
    return LetterType(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
    );
  }
}

class LetterRequest {
  const LetterRequest({
    required this.id,
    required this.letterTypeId,
    required this.letterTypeName,
    required this.purpose,
    required this.status,
    required this.requesterName,
    required this.houseCode,
    required this.householdDisplayName,
    required this.reviewedByName,
    required this.reviewedAt,
    required this.rejectionReason,
    required this.readyAt,
    required this.createdAt,
  });

  final String id;
  final String letterTypeId;
  final String letterTypeName;
  final String purpose;
  final LetterRequestStatus status;
  final String requesterName;
  final String houseCode;
  final String householdDisplayName;
  final String? reviewedByName;
  final DateTime? reviewedAt;
  final String? rejectionReason;
  final DateTime? readyAt;
  final DateTime createdAt;

  factory LetterRequest.fromJson(Map<String, dynamic> json) {
    final reviewedAt = json['reviewedAt'] as String?;
    final readyAt = json['readyAt'] as String?;
    return LetterRequest(
      id: json['id'] as String,
      letterTypeId: json['letterTypeId'] as String,
      letterTypeName: json['letterTypeName'] as String,
      purpose: json['purpose'] as String,
      status: letterRequestStatusFromApi(json['status']),
      requesterName: json['requesterName'] as String,
      houseCode: json['houseCode'] as String,
      householdDisplayName: json['householdDisplayName'] as String,
      reviewedByName: json['reviewedByName'] as String?,
      reviewedAt: reviewedAt != null ? DateTime.parse(reviewedAt) : null,
      rejectionReason: json['rejectionReason'] as String?,
      readyAt: readyAt != null ? DateTime.parse(readyAt) : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}
